build/build-components/build-components.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2019, Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Script to build BL2, BL3 and fitImage components.
+
+The script uses the bitbake_util module to run custom commands in a
+preexistent bitbake environemnt.
+"""
+import sys
+import argparse
+
+import file_util
+from bitbake_util import Bitbake
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--builddir",
+        metavar="DIR",
+        type=file_util.str_to_resolved_path,
+        help="directory in which to build",
+        required=True,
+    )
+    parser.add_argument(
+        "--outputdir",
+        metavar="PATH",
+        type=file_util.str_to_resolved_path,
+        help="directory in which to place build artifacts",
+        required=True,
+    )
+    parser.add_argument(
+        "--machine", metavar="STRING", help="Machine to build.", required=True
+    )
+    parser.add_argument(
+        "--distro",
+        metavar="STRING",
+        help="Name of the distro to build.",
+        default="mbl-development",
+        required=False,
+    )
+    parser.add_argument(
+        "--parent-command-line",
+        metavar="STRING",
+        help=(
+            "Specify the command line that was used to invoke the"
+            "script that invokes build.sh."
+        ),
+        required=False,
+    )
+    parser.add_argument(
+        "--mbl-tools-version",
+        metavar="STRING",
+        help="Specify the version of mbl-tools that this script came from.",
+        required=False,
+    )
+
+    args = parser.parse_args()
+    file_util.ensure_is_directory(args.builddir)
+    return args
+
+
+def main():
+    """Script entry point."""
+    args = _parse_args()
+
+    # Set tup the Bitbake environemnt
+    bitbake = Bitbake(
+        builddir=args.builddir,
+        outputdir=args.outputdir,
+        machine=args.machine,
+        distro=args.distro,
+    )
+    bitbake.setup_environment()
+
+    # Build the components
+    bitbake_build_commands = [
```

Comment:
You have to clean the shared state for the `virtual/bootloader` and `optee-os` as well.
I suggest you construct the command like this:
```
[ "bitbake -c cleansstate virtual/atf optee-os virtual/bootloader virtual/kernel",
"bitbake image_place_holder",
]
```

The `image_place_holder` is the image we want to build (mbl-image-development or mbl-image-production depending on the --distro used). Calling the bitbake to re-generate the image guarantees that all software components that were cleaned will be rebuilt and the components images and signing will be re-processed.

Another important aspect is if we cleansstate these components a new key for the FIP and FIT image will be used, which will break the update, unless we save them in the bitbake ${TOPDIR} directory.

**NOTE:** for rpi3 the b1 and kernel components are packaged together.